### PR TITLE
tap factory unit tests

### PR DESCRIPTION
**Pull request by cgardens**
## What
* Unit test for SingerTapFactory
* Move a bunch of test config creation to a helper.

## Recommended reading order
1. `dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java`
1. the rest

### CODE REVIEW

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    StandardDiscoverSchemaInput discoverSchemaInput = new StandardDiscoverSchemaInput();
+    discoverSchemaInput.setConnectionConfiguration(
+        tapConfig.getSourceConnectionImplementation().getConfiguration());
+    ProcessBuilderFactory pbf = mock(ProcessBuilderFactory.class);
+    ProcessBuilder processBuilder = mock(ProcessBuilder.class);
+    Process process = mock(Process.class);
+    InputStream inputStream = mock(InputStream.class);
+    SingerDiscoverSchemaWorker discoverSchemaWorker = mock(SingerDiscoverSchemaWorker.class);
+
+    final SingerColumn singerColumn = new SingerColumn();
```
Comment (cgardens) [inline dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:110]: i'll move all of this into a helper on the next test suite as i think i will have to reuse it.

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            "--properties",
+            CATALOG_JSON_FILENAME,
+            "--state",
+            STATE_JSON_FILENAME))
+        .thenReturn(processBuilder);
+    when(processBuilder.redirectError(errorLogPath.toFile())).thenReturn(processBuilder);
+    when(processBuilder.start()).thenReturn(process);
+    when(process.getInputStream()).thenReturn(inputStream);
+
+    when(streamFactory.create(any())).thenReturn(expected.stream());
```
Comment (cgardens) [inline dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:149]: i don't like the `any()` here but i don't think there's a reasonable equality i can use on `BufferedReader` to get mockito to trigger this on a more tightly specified argument.

File: dataline-workers/src/main/java/io/dataline/workers/StreamFactory.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+ * SOFTWARE.
+ */
+
+package io.dataline.workers;
+
+import io.dataline.config.SingerMessage;
+import java.io.BufferedReader;
+import java.util.stream.Stream;
+
+public interface StreamFactory {
```
Comment (cgardens) [inline dataline-workers/src/main/java/io/dataline/workers/StreamFactory.java:31]: this is a bad name. suggestions?

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    final SingerTapFactory tapFactory =
+        new SingerTapFactory(IMAGE_NAME, pbf, streamFactory, discoverSchemaWorker);
+    final Stream<SingerMessage> actual = tapFactory.create(tapConfig, jobRoot);
+
+    assertTrue(Files.exists(jobRoot));
+    assertTrue(Files.exists(jobRoot.resolve("tap_config.json")));
+    assertTrue(Files.exists(jobRoot.resolve("catalog.json")));
+    assertTrue(Files.exists(jobRoot.resolve("input_state.json")));
```
Comment (cgardens) [inline dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:158]: i should check that the actual contents of these files are what i expect them to be.

File: dataline-workers/src/main/java/io/dataline/workers/StreamFactory.java (hunk 3 above)
Comment (michel-tricot) [reply dataline-workers/src/main/java/io/dataline/workers/StreamFactory.java:31]: It's ok :)

File: dataline-workers/src/main/java/io/dataline/workers/protocol/singer/SingerJsonStreamFactory.java [hunk 5]
```diff
@@ -42,7 +43,7 @@
  * SingerMessage will still be parsed. If there are multiple SingerMessage records on the same line,
  * only the first will be parsed.
  */
-public class SingerJsonStreamFactory {
+public class SingerJsonStreamFactory implements StreamFactory {
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/protocol/singer/SingerJsonStreamFactory.java:46]: Can you update the code to not use `ObjectMapper`?

File: dataline-workers/src/main/java/io/dataline/workers/singer/SingerTapFactory.java [hunk 6]
```diff
@@ -135,8 +152,7 @@ public Runnable getCloseFunction() {
     StandardDiscoverSchemaInput discoveryInput = new StandardDiscoverSchemaInput();
     discoveryInput.setConnectionConfiguration(
         input.getSourceConnectionImplementation().getConfiguration());
-    Path scopedWorkspace = workspaceRoot.resolve("discovery");
-    return new SingerDiscoverSchemaWorker(imageName, pbf)
-        .runInternal(discoveryInput, scopedWorkspace);
+    Path scopedWorkspace = workspaceRoot.resolve(DISCOVERY_DIR);
```
Comment (michel-tricot) [inline dataline-workers/src/main/java/io/dataline/workers/singer/SingerTapFactory.java:155]: can you rename `workspaceRoot` to `jobRoot`? (I missed that spot. the idea is workspace is the root dir and job root is the scoped dir)

File: dataline-workers/src/test/java/io/dataline/workers/TestConfigHelpers.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  public static ImmutablePair<StandardSync, StandardSyncInput> createSyncConfig() {
+    final UUID workspaceId = UUID.randomUUID();
+    final UUID sourceSpecificationId = UUID.randomUUID();
+    final UUID sourceImplementationId = UUID.randomUUID();
+    final UUID destinationSpecificationId = UUID.randomUUID();
+    final UUID destinationImplementationId = UUID.randomUUID();
+    final UUID connectionId = UUID.randomUUID();
+
+    final String sourceConnection =
+        JsonUtils.toJson(
```
Comment (michel-tricot) [inline dataline-workers/src/test/java/io/dataline/workers/TestConfigHelpers.java:57]: You need to get master into your branch. JsonUtils doesn't exist anymore. `Jsons.serialize(xxx)`

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+  private Path jobRoot;
+  private Path discoverSchemaJobRoot;
+  private Path errorLogPath;
+
+  @BeforeEach
+  public void setup() {
+    try {
+      jobRoot = Files.createTempDirectory(JOB_ROOT_PREFIX);
+      discoverSchemaJobRoot = jobRoot.resolve(DISCOVER_SCHEMA_JOB_ROOT_PREFIX);
```
Comment (michel-tricot) [inline dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:81]: shouldn't you be using the static defined in the source class?

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class SingerTapFactoryTest {
+  private static final String IMAGE_NAME = "hudi:latest";
+  private static final String JOB_ROOT_PREFIX = "workspace";
+  private static final String DISCOVER_SCHEMA_JOB_ROOT_PREFIX = "discover";
```
Comment (michel-tricot) [inline dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:66]: Why do you need to redefine these statics?

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  private Path errorLogPath;
+
+  @BeforeEach
+  public void setup() {
+    try {
+      jobRoot = Files.createTempDirectory(JOB_ROOT_PREFIX);
+      discoverSchemaJobRoot = jobRoot.resolve(DISCOVER_SCHEMA_JOB_ROOT_PREFIX);
+      errorLogPath = jobRoot.resolve(DefaultSyncWorker.TAP_ERR_LOG);
+
+    } catch (IOException e) {
```
Comment (michel-tricot) [inline dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:84]: Don't catch in tests, it makes the test unnecessarily verbose, just add throws to the prototype

**Review by michel-tricot — APPROVED** [review]

File: dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java (hunk 9 above)
Comment (cgardens) [reply dataline-workers/src/test/java/io/dataline/workers/singer/SingerTapFactoryTest.java:66]: :+1: used the static that's defined in the main class for discover. the job root for this class is just any string from its point of view, so leaving that one.

**Pull request merged**